### ISSUE TITLE
The GIT_CONFIG environment variable is not honored

### ISSUE THREAD

**Issue by ararslan** [libgit2/libgit2#4815]
The [`GIT_CONFIG`](https://git-scm.com/docs/git-config#_environment) environment variable tells Git where to look for a configuration file to use. If left unspecified, the default is used. It seems this variable is ignored, be it intentionally or otherwise, by libgit2.

### Reproduction steps

Create a `.gitconfig` file in some directory, and specify user information. For example,
```ini
[user]
    name = Definitely Not Alex
    email = something@else.gov
```
Now set an environment variable `GIT_CONFIG` containing a path to this file. Compile and run this C code, making sure you're inside a Git repo:
```c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <git2.h>

int main(int argc, char** argv) {
    printf("GIT_CONFIG: %s\n", getenv("GIT_CONFIG"));

    git_libgit2_init();

    git_repository* repo = NULL;
    char wd[1024];
    getcwd(wd, sizeof(wd));
    int error = git_repository_open(&repo, wd);

    git_config* cfg = NULL;
    error = git_repository_config(&cfg, repo);

    const git_config_entry* entry = NULL;
    error = git_config_get_entry(&entry, cfg, "user.name");
    printf("libgit2 config username: %s\n", entry->value);

    git_config_entry_free(entry);
    git_config_free(cfg);
    git_repository_free(repo);
    git_libgit2_shutdown();
    return 0;
}
```

### Expected behavior

The configuration as specified in the given file should be used.
```
$ GIT_CONFIG=/Users/alex/testing/.gitconfig ./test
GIT_CONFIG: /Users/alex/testing/.gitconfig
libgit2 config username: Definitely Not Alex
```

### Actual behavior

The default global configuration is used.
```
$ GIT_CONFIG=/Users/alex/testing/.gitconfig ./test
GIT_CONFIG: /Users/alex/testing/.gitconfig
libgit2 config username: Alex Arslan
```

### Version of libgit2 (release number or SHA1)

0.27.4

### Operating system(s) tested

macOS High Sierra

**Comment by neithernut** [comment]
See https://github.com/libgit2/libgit2/issues/2910#issuecomment-385339402: libgit2 (and any other sane library) will not change its behavior based on environment variables (except e.g. `LANG`). It is the responsibility of the application's author to perform/choose a configuration sane for the use-case.

In the case of a "git-plugin", this would mean honoring `GIT_CONFIG`. In the case of a web-service, things tend to be different.

Maybe this could be stated in the documentation more clearly, maybe a convenience version of `git_repository_config()` could be added (`git_repository_config_for_apps()` or something).

**Comment by ethomson** [comment]
I think it might make sense for `git_repository_open_ext` to honor this when `GIT_REPOSITORY_OPEN_FROM_ENV` is set.

**Comment by neithernut** [comment]
Mh... `git_repository_open_ext()` already offers some convenience functionality: if called with the flag `GIT_REPOSITORY_OPEN_FROM_ENV`, it already resorts to environment variables for finding the repository. Controlling how to look for the configuration could be seen as an extension. So yeah, it sounds reasonable.

However, imho whether `GIT_CONFIG` is honored or not should be controllable via a flag (e.g. `GIT_REPOSITORY_OPEN_CONFIG_FROM_ENV`). Bundling the behavior with `GIT_REPOSITORY_OPEN_FROM_ENV` would be wrong since the semantics are different. Having no control over whether the environment is used for the config or not wouldn't be too great either.

**Comment by pks-t** [comment]
I agree with you, @neithernut. Introducing a separate flag to `open_ext` is the way to go in my opinion.

**Comment by carlosmn** [comment]
I'm with @ethomson on this. The `_FROM_ENV` option already indicates that we want to open the repository like git would given the current environment variables. This fits right in with it.

**Comment by tiennou** [comment]
I'm also tempted to side with @ethomson : if you want to respect the environment, use that. If you don't like the environment, just `unsetenv` the parts you don't want before calling that, or maybe even don't set the environment in this way. Arguably, the "client" might not be in control of it, but IMHO that's a problem with anything touching the environment (or running external executables, for that matter).

That said, the cursory look I gave at handling `GIT_CONFIG` entails keeping track of a repository having been loaded from the env, so that reloading its configuration doesn't accidentally reset the repo to the normal one (and it seemed convoluted).

**Comment by pks-t** [comment]
If you ask me, then this would be changing the API in much too subtle ways. And this is the worst kind of API change: it's not detected by the compiler, but it may still change how your program behaves in unexpected ways because you have previously not been told that this may happen. 

The documentation of `GIT_REPOSITORY_OPEN_FROM_ENV` explicitly mentions all environment variables that impact the way the function behaves. So a user of that option may carefully have sanitized the environment to respect only a subset of them. If we now just add a new environment variable that he'll have to sanitize, then this breaks his expectations from the API and the developer will probably never ever notice this because the code continues to compile just fine.

Furthermore, it wouldn't give developers the ability to choose. The option is about opening the _repository_ from the environment, not the configuration. While these are usually heavily tied together, a developer might have reasons to want one, but not the other. He might wish to open configurations from the environment, but just take the repo at the specified path or vice versa. If we tie those two unrelated things together, then the developer isn't able to do this anymore.

**Comment by tiennou** [comment]
My initial POV was to consider failing to account for `GIT_CONFIG` a bug of `_FROM_ENV`, as its documentation (also) explicitly says :

> Find and open a git repository, respecting the environment variables used by the git command-line tools. But the API change is a good point, so I'd like to back up a bit and ponder.

For reference, The Book has a [list](https://git-scm.com/book/uz/v2/Git-Internals-Environment-Variables) of environment variables "used" by `git` (I realize that `GIT_CONFIG` isn't actually documented there 🤷‍♂️).

Grepping the `git` source code gives this :
- `GIT_CONFIG` (`#defined` as `CONFIG_ENVIRONMENT`), is only used by the `git-config` tool.
- there's a list of "local" env vars in [environment.c](https://github.com/git/git/blob/master/environment.c#L115), that makes apparent whose envvars can be used.
- the release notes from [1.6.0](https://github.com/git/git/blob/master/Documentation/RelNotes/1.6.0.txt#L41) say this :

> GIT_CONFIG, which was only documented as affecting "git config", but
actually affected all git commands, now only affects "git config".
GIT_LOCAL_CONFIG, also only documented as affecting "git config" and
not different from GIT_CONFIG in a useful way, is removed.

I… don't think I understand what it means, since `git` involves both shell scripts (likely to go through `git-config` anyway) and functions calls (which AFAICS don't seem to use the envvar), I'd be tempted to rephrase as "internal accesses (function calls) to the config never use `GIT_CONFIG`, while external/shell script access do". If that's wrong most of what I'm about to say is too 😉.

It means we must add a new "constructor" to `git_config`, that accounts for the env var. Everything else stays as it is. Doesn't look useful though, but at least the low-level code is done 😉.

Now, about that API compatibility w.r.t envvars that are (especially envvars that will be), I'd do the following :

- add `git_repository_open_with_opts`, which takes a struct made up of the current `git_repository_open_ext` parameters.
- add a `git_strarray allowed_envvars` to the struct, in effect whitelisting what envvars we're allowed to use. `_FROM_ENV` becomes "use the envvars in `allowed_envvars` or all of them if it's empty`.
- freeze the current list of envvars in `git_repository_open_ext`, so we don't accidentally start using an unwhitelisted envvar when support is added.

This allows old code to pick and choose what they want, at their leisure, if they want specific filtering, while more trusty ones would get instant ~surprises~ support.

With that, we can add support for any envvar we want without thinking about backward-compatibility (and without growing `git_repository_open_flag_t` with every envvar we stumble on).

One obvious limitation is that "systems" effectively need access to the whitelist/the repo (a contrived example would be `GIT_GLOB_PATHSPECS`). An alternative would be to make that a library option.

Does that sound like a plan ?

**Comment by ethomson** [comment]
So...  in retrospect, yes, `git_repository_open_ext` should have taken an extensible options structure.  My bad.  😥

In any case, I think that your plan makes sense if we really need to create a version of `git_repository_open` with an extension option structure.  And we could do that to future-proof this... but I think that we could also just add a `GIT_REPOSITORY_CONFIG_FROM_ENV` to handle this request.

(Of course, I'm also in the `GIT_REPOSITORY_OPEN_FROM_ENV` should have handled this case and we should fix it camp.)

**Comment by ethomson** [comment]
See #3854 as well

**Comment by pks-t** [comment]
Your proposal sounds perfectly reasonable to me, @tiennou
